src/mpi/coll/algorithms/common/coll_progress_util.h

```diff
@@ -0,0 +1,94 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+#ifndef MPIR_ALGO_NAMESPACE
+#error "The template must be namespaced with MPIR_ALGO_NAMESPACE"
+#endif
+
+/*FIXME: MPIR_COLL_progress_global should be per-VNI */
+extern MPIR_COLL_progress_global_t MPIR_COLL_progress_global;
+
+#undef FUNCNAME
+#define FUNCNAME MPIR_ALGO_sched_wait
+/* Routing to kick-start and completely execute a blocking schedule*/
+MPL_STATIC_INLINE_PREFIX void MPIR_ALGO_sched_wait(MPIR_TSP_sched_t * sched)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_ALGO_SCHED_WAIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_ALGO_SCHED_WAIT);
+
+    int rc = 0;
+
+    while (!rc) {
+        rc = MPIR_TSP_poke(sched);   /* Make progress on the collective schedule */
+        if (!rc) {
+            MPIR_COLL_progress_global.progress_fn();
+        }
+    }
+
+    /* Mark sched as completed */
+    MPIR_TSP_sched_finalize(sched);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_ALGO_SCHED_WAIT);
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIR_ALGO_kick_nb
+/* Function to make progress on a non-blocking collective*/
+MPL_STATIC_INLINE_PREFIX int MPIR_ALGO_kick_nb(MPIR_COLL_queue_elem_t * elem)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_ALGO_KICK_NB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_ALGO_KICK_NB);
+
+    int done = 0;
+    MPIR_TSP_sched_t *sched = (MPIR_TSP_sched_t *) (((MPIR_ALGO_req_t *) elem)->sched);
+
+    /* make progress on the collective */
+    done = MPIR_TSP_poke(sched);
+
+    if (done) {
+        /* remove the collective from the queue */
+        TAILQ_REMOVE(&MPIR_COLL_progress_global.head, elem, list_data);
+
+        /* delete all memory associated with the schedule */
+        MPIR_TSP_sched_finalize(sched);
+        ((MPIR_ALGO_req_t *) elem)->sched = NULL;
+    }
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_ALGO_KICK_NB);
+
+    return done;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIR_ALGO_sched_test
+/* Routing to kick-start a non-blocking schedule*/
+MPL_STATIC_INLINE_PREFIX int MPIR_ALGO_sched_test(MPIR_TSP_sched_t * sched, MPIR_ALGO_req_t * request)
```

Comment:
I have to change the name of this function to MPIR_ALGO_sched_append_and_poke.

Reply:
That makes a little more sense, but I'm still not sure how this is useful. We could either make all appends poke the progress engine or have an explicit wait later (or both).

Reply:
This function is used for non-blocking collectives. When a non-blocking collective is called, it's schedule is appended to the non-blocking collectives schedule queue and the schedule is poked to start its execution. This is what this function does. Later on when wait is called on that collective, there is a progress function to make progress. I don't think I understand your concern here.

Reply:
I'm just not sure why we don't either do:

1. _Always_ append and poke the progress engine to move things along if that's what we want, even with the blocking collectives.

or

2. Schedule everything and then wait for it later.

It seems like this is an unnecessary optimization that I'm not sure actually optimizes anything. I might be wrong, and if so I'd like to see it. 😄

Reply:
For blocking collectives, we don't need to go to the progress engine. The transport scheduler is sufficient to make sure that all the tasks in the schedule are executed. 
For non-blocking collectives, when the collective is called, the schedule is first generated and then appended to the progress engine. Whenever the control goes to MPI, it can make progress on those schedules by calling the corresponding transport scheduler. 

We can't schedule everything because there are dependencies across tasks and the transport scheduler makes sure that the tasks are executed only when the dependencies are satisfied.

Also this function need not be a MPIR_ALGO function. This could just be a MPIR_TSP function. We can talk to clarify, we might be confusing the terminologies being used here.